components/builder-sessionsrv/src/server/handlers.rs

```diff
@@ -106,7 +105,18 @@ pub fn session_create(
     state: &mut ServerState,
 ) -> SrvResult<()> {
     let mut msg = req.parse::<proto::SessionCreate>()?;
-    let mut session = Session::default();
+    debug!("session-create, {:?}", msg);
+    {
+        if let Some(session) = state.sessions.read().unwrap().get(&(
+            msg.get_extern_id(),
+            msg.get_provider(),
+        ))
+        {
+            debug!("found existing session, {:?}", session);
+            conn.route_reply(req, &**session)?;
+            return Ok(());
+        }
+    }
```

Comment:
Why is this chunk of code in its own block? Is it for interacting with the `state`?

Reply:
yeah probably to not hold on to the read lock on sessions..

Reply:
Yep, it's because we're entering a critical section of memory and want to release it ASAP